### PR TITLE
process_unix: prefer i32::*_be_bytes over manually shifting bytes

### PR DESCRIPTION
**Pull request by tesuji**
This PR makes it more clear about the intend of the code.

### CODE REVIEW

File: src/libstd/sys/unix/process/process_unix.rs [hunk 1]
```diff
@@ -52,11 +53,12 @@ impl Command {
...
+                    let errno = errno.to_be_bytes();
                     let bytes = [
-                        (errno >> 24) as u8,
-                        (errno >> 16) as u8,
-                        (errno >> 8) as u8,
-                        (errno >> 0) as u8,
+                        errno[0],
+                        errno[1],
+                        errno[2],
+                        errno[3],
```
Comment (LukasKalbertodt) [inline src/libstd/sys/unix/process/process_unix.rs:61]: Uhm... with this code, the `errno` is always encoded in big endian into the array. But before, it was always swapped -- even on big endian devices -- right? And I don't see any `#[cfg(target_endianess = ...)]` attributes anywhere. So your PR changes behavior, right? Which means either your change or the previous code is buggy? Is this code tested on big endian CPUs?

I unfortunately don't know a lot about these unix system APIs. Could you maybe link me some resource which specifies that this has to be big endian? Or is this *just* used internally? Like... the data is used below. But if it's just internally, why swap any bytes at all? Why not keep it in native endianess?

File: src/libstd/sys/unix/process/process_unix.rs [hunk 2]
```diff
@@ -103,13 +106,8 @@ impl Command {
...
 
-        fn combine(arr: &[u8]) -> i32 {
-            let a = arr[0] as u32;
-            let b = arr[1] as u32;
-            let c = arr[2] as u32;
-            let d = arr[3] as u32;
-
-            ((a << 24) | (b << 16) | (c << 8) | (d << 0)) as i32
+        fn combine(arr: [u8; 4]) -> i32 {
+            i32::from_be_bytes(arr)
```
Comment (LukasKalbertodt) [inline src/libstd/sys/unix/process/process_unix.rs:110]: Same as above: before it always swapped, now it only swaps on little endian system.

**Review by LukasKalbertodt — COMMENTED** [review]
Unfortunately, I am somewhat confused by the PR :/ 
I am just very unfamiliar with that code. Could you just link me to some resources that explain the unix basics required for understanding this code? I mean, I wouldn't have to fully understand it if it weren't for that fact that, to me, it looks like your change is changing behavior.

File: src/libstd/sys/unix/process/process_unix.rs (hunk 2 above)
Comment (lzutao) [reply src/libstd/sys/unix/process/process_unix.rs:110]: The generated asms are the same: https://rust.godbolt.org/z/ezbn7K .
IIUC, the old code is loading Big Endian bytes from `arr`. 
arr[0] is most significant value, arr[3] is the least significant one.

**Comment by lzutao** [comment]
You could see the introduction to Endianness in [wiki](https://en.wikipedia.org/wiki/Endianness) or [boost](https://www.boost.org/doc/libs/1_72_0/libs/endian/doc/html/endian.html) .

**Review by LukasKalbertodt — APPROVED** [review]
Thanks for your links, but I *do* know what endianness is :smile: 
I was rather asking for docs on why a specific endianness is required in this particular case/code (dealing with unix APIs).

*However*, I have been bamboozled by the code! I misunderstood the original code and therefore was very confused! Now everything makes sense again and I like that change!

File: src/libstd/sys/unix/process/process_unix.rs [hunk 3]
```diff
@@ -81,12 +83,13 @@ impl Command {
             match input.read(&mut bytes) {
                 Ok(0) => return Ok((p, ours)),
                 Ok(8) => {
+                    let (errno, footer) = bytes.split_at(4);
                     assert!(
-                        combine(CLOEXEC_MSG_FOOTER) == combine(&bytes[4..8]),
+                        combine(CLOEXEC_MSG_FOOTER) == combine(footer.try_into().unwrap()),
```
Comment (lzutao) [inline src/libstd/sys/unix/process/process_unix.rs:88]: Actually, why we not just compare 2 slice instead?

Comment (LukasKalbertodt) [reply src/libstd/sys/unix/process/process_unix.rs:88]: Oops, that would be better indeed.

**Pull request merged**